# Part 1: Collecting subreddit "India" Data

Source: https://towardsdatascience.com/scraping-reddit-data-1c0af3040768

Importing necessary libraries

In [1]:
import pandas as pd # data manipulation 
import praw # python reddit API wrapper

/home/prakhar/.local/lib/python3.6/site-packages/pandas/compat/__init__.py:117: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.
  warnings.warn(msg)


Authenticate ourselves by creating a Reddit instance and providing client credentials

In [2]:
# Credentials generated from the reddit developers applications page
my_client_id = '8KS6G6Nt9BU9sg'
my_client_secret = 'CkeVFda-vf0DbseDb0eEr1YMpJo'
user = 'reddit_scrape'

reddit = praw.Reddit(client_id=my_client_id, client_secret=my_client_secret, user_agent=user)

## Get subreddit data 
Using the `reddit` instance from the previous section, we can aquire top 1000 posts or 1000 hottest posts or latest posts from reddit. 

In [3]:
num_of_posts = 1000   # Number of posts we want in our data
new_posts = reddit.subreddit('India').top(limit=num_of_posts)

In [4]:
# Printing the titles of the top_posts in this subreddit
for post in new_posts:
    print(post.title)

Will donate thrice the number of upvotes (amount in Rs.) i get for this thread in next 24 hours
Indian reply to NYtimes cartoon on Paris climate accord by Satish Acharya.
The essence of the Indian soap opera, distilled into one GIF.
Fuck all Religion
German exchange Student at IIT Madras is being sent back home by the Indian immigration department because he joined the protest.
Tragedy of India
Today's The Hindu
If you are not moved by this picture, I wish I had your heart. [NP]
Megathread: India-Pakistan border skirmish
"From midnight the entire country will go under a complete lockdown," says PM Modi.
The wealth inequality in India is truly horrifying
Dear Hindus, we Indian Muslims rejected an Islamic state in 1947.. Now it's your turn to reject a Hindu Nazi state.. Speak up. NOW
Spend nearly 500 rs bucks to watch JOKER and in return all i got was 30 mins of advertiment before the movies,20 mins adver. in interval, giant subtitles, and stupid tobacco discalimer which stuck through 70

An ode to Gandhi ( from twitter )
Every home today morning
Need all the help possible.
Found at a hospital in India.
Bengaluru City Police
Pretty much the current state of India
Just a photo of HP Petrol Pump I took recently. The line drawing is added in PS later. [OC] 3922x4902
[OC] This Contest From Maggi.
An in depth news coverage of the monsoons
Modern day mathematics for modern problems...Brilliant stuff!
#satisfying
This is too funny
Pic From Rural Kerala
The Times Of Modi!
Science students can relate
I am a Muslim and I wish all my Hindu brothers and sisters a happy Dusshera!
YES Bank
It's confirmed but also not confirmed.
The Original 10-Year Challenge
When Google gives up
A Big Lake in Chennai - 2018 vs. 2019
Something I made given the time _/\_
Good boi in traffic
At the Pune protest at Pune University
This girl with a "Donate a book to the needy" placard
First Mango 🥭 in our home
Chill!
Congratulations to the Indian Women's Rugby Team who have scripted history by registering

Indian auto drivers during a traffic jam
Translation: Be careful. We ought to save ourselves too while we save them.
I saw your Meghalya post and i raise you this boi.... Someplace on the way to Kheerganga, HP.
Happy birthday SIR APJ ABDUL KALAM. The missile man would've been 88 today.
Viral Hindu-Muslim, India-Pak lesbian couple celebrates anniversary with new pics. They are stunning
Kolkata at night during Durga Pujo
For all our desi Calvin and Hobbes and Chacha Chaudhary fans!
LED traffic strip lights in Hyderabad
I made 13 illegal toll booth vanish overnight, milking 3-4 lakhs everyday, using RTI
Deccan Herald's Speak Out: 7th March 2019.
The 90's is (calcium) strong with this one!
When you have more phones than toilets
This year's hand made Ganesha by my sis
[Showerthought] Netflix should provide a "Skip Song" Button for Indian movies.
I see your good boi in traffic with purrtty kitty in traffic.
Rahul Dravid "I don't think I can just get a doctorate because I happen to be a famou

I work as a psychologist for children and teens. This is from a 10th standard student I've been counselling for the past few years.
No 'Clapping' Doesn't Kill Coronavirus.. God help this society.
This pic of Mumbai Conservancy workers cleaning gutters.....
We've Just Seen the First Use of Deepfakes in an Indian Election Campaign
Cartoon of the day: Jharkhand Election Results by @CartoonistAlok
Kiki challenge. Indian edition
They're just everywhere these days.
My neighbour's door, summarises cultural values of India !!
How do you know if someone went to IIT? They'll tell you.
I finally convinced him. That awesome moment when you understand the essense of life
1981: When Atal Bihari Vajpayee broke protocol to meet Faiz Ahmed Faiz.
Hasan Minhaj Has Heard His Fans, Next 'Patriot Act' Episode on CAA-NRC
Srinivasa Gowda (28) from Moodabidri in Dakshina Kannada district, Karnataka who Ran 142.5 meters in just 13.62 seconds at a "Kambala" or Buffalo race in a slushy paddy field. 100 meters in 

Not in my name! #Lynching #JaiShriRam
A normal day in nagpur.
Choose one.
'Triple Talaq Unconstitutional', Says Supreme Court In Majority Verdict
GTA 5 girl wearing a beautiful saree
Munnar, Kerala
As in depth news coverage of the monsoons (Repost) [NP]
Johnny Sins, Mia Khalifa and Sunny Leone welcome you to Sree Narayana Polytechnic College, Kottiyam, Kerala.
Aizawl has these tiny roadside libraries
My uber driver last night, Rakesh, carries his own water from home in a thermos and a cup, do reduce waste generated by bottles.
Breaking news: Sridevi passes away
Hi I'm Shashi Tharoor. Ask me Anything on India, politics, foreign affairs, history, and more!
Preach!🙌
[OC] I see your Milky Way from the Himalayas and Mumbai. I raise you this shot from Bhopal. Comparison between one exposure vs 120 exposures stacked together. [Nikon D3100, 2608x3509] (X-post from r/space)
Farmer Gets Rs 1,064 For 750 kg Of Onion, Sends Money To PM In Protest
Some people never learn from their past mistakes.
H

Looking at the data type of the `new_posts` to identify the class it belongs to and to get more information about the various methods and attributes. 
Answer - object of the ListingGenerator class

In [5]:
print(type(new_posts))

<class 'praw.models.listing.generator.ListingGenerator'>


## Structure subreddit data into a pandas dataframe

Create a list of features that have to be saved in the dataset. These are data columns that we will extract from our dataset. 

**Feature Description:**

* **Title:** The title of the submission.
* **Score:** The number of upvotes for the submission.
* **ID:** ID of the submission.
* **Subreddit:** Provides an instance of Subreddit.
* **URL:** The URL the submission links to, or the permalink if a selfpost.
* **Orginal**: Whether or not the submission has been set as original content.
* **num_comments:** The number of comments on the submission.
* **Body:** The submissions’ selftext - an empty string if a link post.
* **created_on:** Time the submission was created, 


In [6]:
features = ['Title', 'Score', 'ID', 'Subreddit', 'URL', 'Original', 'num_comments', 'Flair', 'Body', 'created_on']
posts = [] # List containing the data from individual reddit posts. Each item will be a new entry in the dataframe

In [7]:
india_sub = reddit.subreddit('India')

# Loop through each subreddit entry and append that t
for post in india_sub.top(limit=num_of_posts):
    posts.append([post.title, post.score, post.id, post.subreddit, post.url, post.is_original_content, post.num_comments, post.link_flair_text, post.selftext, post.created])

# Convert to a data frame 
posts = pd.DataFrame(posts, columns=features)

Let's look at the data frame and analyse the data a little more. 

In [8]:
posts

Title  Score      ID  \
0    Will donate thrice the number of upvotes (amou...  19710  981o7s   
1    Indian reply to NYtimes cartoon on Paris clima...  18246  6f10op   
2    The essence of the Indian soap opera, distille...  18199  8pymkp   
3                                    Fuck all Religion  17902  f9outu   
4    German exchange Student at IIT Madras is being...  11726  eev8g5   
..                                                 ...    ...     ...   
983  GSLV Mk 3 launch as seen from my office in Che...   1412  6fduxo   
984                                Slayed by DH, again   1414  a47h3n   
985  Taken a while back from the PSLV satellite lau...   1415  brj600   
986  Mods are sleeping, upvote this lotus to the fr...   1412  2uryz4   
987                      Morning Scenes in my village!   1409  8hkla8   

    Subreddit                                                URL  Original  \
0       india  https://www.reddit.com/r/india/comments/981o7s...     False   
1       india                           http://imgur.com/a/U48v9     False   
2       india           https://gfycat.com/DigitalSparseAkitainu     False   
3       india  https://www.reddit.com/r/india/comments/f9outu...     False   
4       india                https://i.redd.it/j816l4c40i641.jpg     False   
..        ...                                                ...       ...   
983     india           https://gfycat.com/HideousHospitableLamb     False   
984     india                https://i.redd.it/1vvsdfh2bz221.jpg     False   
985     india                https://i.redd.it/xo70b4wz9oz21.jpg      True   
986     india                     http://i.imgur.com/qg1bzHZ.jpg     False   
987     india                https://i.redd.it/vroh12vqrcw01.jpg     False   

     num_comments               Flair  \
0             831       [R]eddiquette   
1            1492              /r/all   
2             946               r/all   
3            4192            Politics   
4             501            Politics   
..            ...                 ...   
983           110  Science/Technology   
984            68            Politics   
985            59         Photography   
986           128            Politics   
987            79       Non-Political   

                                                  Body    created_on  
0    >**Note**: If you want to know what this is al...  1.534536e+09  
1                                                       1.496523e+09  
2                                                       1.528635e+09  
3    Fuck all religion. Fuck Hindusim, fuck Islam, ...  1.582726e+09  
4                                                       1.577186e+09  
..                                                 ...           ...  
983                                                     1.496695e+09  
984                                                     1.544271e+09  
985                                                     1.558520e+09  
986                                                     1.423102e+09  
987                                                     1.525692e+09  

[988 rows x 10 columns]

### Correct the time format using the datetime library

In [9]:
# Using the to_datetime function of pandas
posts['creation_date'] = pd.to_datetime(posts['created_on'], dayfirst=True, unit='s')

In [10]:
# Drop created_on column now 
posts.drop(['created_on'], axis=1, inplace=True)

In [11]:
posts.head()

Title  Score      ID Subreddit  \
0  Will donate thrice the number of upvotes (amou...  19710  981o7s     india   
1  Indian reply to NYtimes cartoon on Paris clima...  18246  6f10op     india   
2  The essence of the Indian soap opera, distille...  18199  8pymkp     india   
3                                  Fuck all Religion  17902  f9outu     india   
4  German exchange Student at IIT Madras is being...  11726  eev8g5     india   

                                                 URL  Original  num_comments  \
0  https://www.reddit.com/r/india/comments/981o7s...     False           831   
1                           http://imgur.com/a/U48v9     False          1492   
2           https://gfycat.com/DigitalSparseAkitainu     False           946   
3  https://www.reddit.com/r/india/comments/f9outu...     False          4192   
4                https://i.redd.it/j816l4c40i641.jpg     False           501   

           Flair                                               Body  \
0  [R]eddiquette  >**Note**: If you want to know what this is al...   
1         /r/all                                                      
2          r/all                                                      
3       Politics  Fuck all religion. Fuck Hindusim, fuck Islam, ...   
4       Politics                                                      

        creation_date  
0 2018-08-17 20:02:17  
1 2017-06-03 20:47:25  
2 2018-06-10 12:46:29  
3 2020-02-26 14:10:49  
4 2019-12-24 11:10:46

In [12]:
posts.shape

(988, 10)

Even after changing the `num_of_posts`, I still getting only 988 posts and a very skewed dataset. Hence, I need to create a better scraping mechanism to get a more balanced Dataset

# Data Collection: Improvised

The first thing I will do is to create lesser number of targets for classification for more robust models and increase the accuracy. 

In [13]:
posts['Flair'].value_counts().sort_values(ascending=False)

Non-Political                     422
Politics                          295
Coronavirus                        46
Photography                        43
[R]eddiquette                      34
Sports                             29
Business/Finance                   16
Unverified                         15
Policy/Economy                     14
Science/Technology                 14
Food                               13
AskIndia                           10
CAA-NRC                             9
r/all                               3
AMA                                 2
Policy & Economy                    2
/r/all                              2
Demonetization                      1
Misleading Headline                 1
On Internet Shutdowns               1
Totally real                        1
Original Comics                     1
Entertainment                       1
Official Sadness Thread             1
Misleading                          1
Politics [Megathread]               1
Goal Achieve

I sorted the flairs in the descending order and picked up the most popular flairs to avoid skewed data. I will collect top posts and their comments along with some other information relavnt for the analysis and the model. 

In [14]:
# Relevant flairs
flairs = ["AskIndia", "Non-Political", "[R]eddiquette", 
          "Photography", "Science/Technology",
          "Politics", "Business/Finance", "Policy/Economy",
          "Sports", "Food", "AMA", "Coronavirus"]

In [15]:
# Data features that we will be collecting 
features = [
    'Title', 
    'Score', 
    'ID',
    'URL', 
    'num_comments', 
    'created_on', 
    'Body', 
    'Original',
    'Flair', 
    'Comments'
]

In [16]:
# Create a subreddit instance 
subreddit = reddit.subreddit('india')

In [28]:
posts = []
# Top 150 posts of each type 
for flair in flairs: 
    relevant_subs = subreddit.search(f"flair_name:{flair}", limit=150)
    
    for sub in relevant_subs:
        post = []
        post = [
            str(sub.title),
            str(sub.score),
            str(sub.id),
            str(sub.url),
            sub.num_comments,
            str(sub.created),
            str(sub.selftext),
            sub.is_original_content,
            str(sub.link_flair_text),
        ]
        
        sub.comments.replace_more(limit=0)
        comment = ''
        for top_comment in sub.comments:
            comment = str(top_comment.body) + ' '        
        
        post.append(str(comment))# Add to the end of the list 
        posts.append(post)    # Add to the main list 

In [31]:
# Convert to a data frame 
posts = pd.DataFrame(posts, columns=features)

In [32]:
posts

Title Score      ID  \
0     Lost my Job, Sick Mother and Paralysed Dad, In...  1046  g014wc   
1     Why does the government come with a begging bo...   645  fxofyu   
2     Mother's condition is going worse due to hepat...   764  g0zlly   
3     People stuck with their family during the lock...   157  g4lrhm   
4     Men who are 30+ and have decided not to get ma...   267  fvy95j   
...                                                 ...   ...     ...   
1645  IAMA person suffering from Bipola[r] Disorder....    94  2oytx7   
1646                      Identity, policy, and privacy   123  4bmka5   
1647  Hi /r/India, I am cartoonist Sumit Kumar autho...    75  3rhufx   
1648  Hi Reddit, this is XUlrike from Janwaar Castle...    64  4yc03a   
1649  Hey guys, I am Harsh Rajat, started my entrepr...    79  4vs9fj   

                                                    URL  num_comments  \
0     https://www.reddit.com/r/india/comments/g014wc...           134   
1     https://www.reddit.com/r/india/comments/fxofyu...           204   
2     https://www.reddit.com/r/india/comments/g0zlly...            94   
3     https://www.reddit.com/r/india/comments/g4lrhm...           117   
4     https://www.reddit.com/r/india/comments/fvy95j...           206   
...                                                 ...           ...   
1645  https://www.reddit.com/r/india/comments/2oytx7...           274   
1646  https://www.reddit.com/r/india/comments/4bmka5...            80   
1647  https://www.reddit.com/r/india/comments/3rhufx...           159   
1648  https://www.reddit.com/r/india/comments/4yc03a...            68   
1649  https://www.reddit.com/r/india/comments/4vs9fj...            62   

        created_on                                               Body  \
0     1586742148.0  Hi....It's really tough time for everyone. I r...   
1     1586448261.0  We have floods, terrorist attacks, famines due...   
2     1586870771.0  Hi folks, I really appreciate the warm respons...   
3     1587383829.0  I don't think we've spend so much time with fa...   
4     1586206585.0  The corona virus has given me some time to thi...   
...            ...                                                ...   
1645  1418329767.0  Hi all. I alternate between feeling like Einst...   
1646  1458770333.0  Hi Reddit community! It’s a pleasure to be her...   
1647  1446675989.0  Edit : Going to sleep now. Big dhanyawaad for ...   
1648  1471552429.0  The purpose of the Janwaar Castle Community Or...   
1649  1470171020.0  About Me: A little backstory about me: I start...   

      Original     Flair                                           Comments  
0        False  AskIndia  Learn Python, then Django. Php might take a lo...  
1        False  AskIndia                                         [removed]   
2        False  AskIndia  Can I get some updates and verification on thi...  
3        False  AskIndia  >patriarchal father who could care less etc  \...  
4        False  AskIndia  Get married. Indians are genetically engineere...  
...        ...       ...                                                ...  
1645     False       AMA                      Is your zodiac sign Gemini?    
1646     False       AMA                                         [deleted]   
1647     False       AMA  late to the party and i have no questions for ...  
1648     False       AMA  [Your organization’s logo](https://janwaar-cas...  
1649     False       AMA  Okay then! this seems weird, more than that sh...  

[1650 rows x 10 columns]

In [33]:
# More detailed Data
posts.to_csv('data.csv')